## 1. About

This noteboook takes in as input the data gotten from the [SPICE](https://en.wikipedia.org/wiki/SPICE) simulation about the circuit and reformats them into a format more usable.

### 1.1 About this New Version

The old notebook was laggy and not well documented, this notebook aims to improve this. 

Mainly by exchanging the **pandas** libary through **polars** libary.

### 1.2 Explanation of the Data

### 1.3 General

Learnable SG Data

#### Dataset 1

Inputs:  
1. Processed_VIN_of_Constant_1st_Neuron_Vin0-1.csv
2. Processed_VIN_of_Constant_1st_Neuron_Vin1-2.csv

Outputs:  
1. Processed_output_Vout_of_Constant_1st_Neuron_Vin0-1.csv
2. Processed_output_Vout_of_Constant_1st_Neuron_Vin1-2.csv
3. Processed_output_Vout_of_Constant_2nd_Neuron_Vin0-1.csv
4. Processed_output_Vout_of_Constant_2nd_Neuron_Vin1-2.csv

Parameters:     
- a. C1 (10u-100u)
- b. CIN (10u-100u)
- c. COUT (10u-100u)
- d. L (40u-100u)
- e. R1 (20k-50k)
- f. W1 (600u-1000u)
- g. VIN (input voltage) (0:1, 1:2)

#### Dataset 2

Inputs:  
1. Processed_Vin_sin1.csv
2. Processed_Vin_sin2.csv
3. Processed_Vin_sin3.csv

Outputs:  
1. Processed_output_Vout_of_Sin1.csv
2. Processed_output_Vout_of_Sin2.csv
3. Processed_output_Vout_of_Sin3.csv

Parameters:    
- a. C1 (10u-100u)
- b. CIN (10u-100u)
- c. COUT (10u-100u)
- d. L (40u-100u)
- e. R1 (20k-50k)
- f. W1 (600u-1000u)
- g. a  (inputs to create sin wave of multiple freq) (0:0.25, 0.25:0.5,0.5:1)
- h. b (0:180,180:360,180:360)
- i. f (1:3,3:5,3:5)
  

In [1]:
# Number of Entries of the dataset we want to use
# For Testing purpose to generate all the data set to "None"
NUMBER_DATASET_LENGTH = None

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import torch
import re
import warnings
import polars as pl

## 2. Processing Inputs of Dataset 1

We have input voltage data (VIN) in df_input_1 and df_input_2.
And output voltage data (VOUT) in df_output_2 and df_output_2.

Each CSV is organized like this:

```csv
X (time?)	Column 1	Column 2	Column 3	…
0.00	0.04193	0.04193	0.04193	…
0.01	0.04193	0.04193	0.04193	…
```

Columns are named with parameter information inside parentheses, e.g.:

```
/Vout_buffer2 (C1=9.4e-05,CIN=7e-05,COUT=3.5e-05,L=6.6e-05,R1=20473,VIN=1.111111,w1=0.000776)
```

Each column corresponds to a simulation run with a particular set of circuit parameters and input voltage and each row entry represents one timestep, whereby the entry the current input voltage of the time step is.

In [3]:
DATA_LOCATION = "data/V2/"

In [4]:
df_input_1 = pl.read_csv(str(DATA_LOCATION + "Input/" + "Processed_VIN_of_Constant_1st_Neuron_Vin0-1_v2.csv"))
df_input_2 = pl.read_csv(str(DATA_LOCATION + "Input/" + "Processed_VIN_of_Constant_1st_Neuron_Vin1-2_v2.csv"))
df_input_1.head()

shape: (5, 10_000)
┌──────┬────────────┬────────────┬────────────┬───┬────────────┬───────────┬───────────┬───────────┐
│ X    ┆ /VIN (C1=9 ┆ /VIN (C1=9 ┆ /VIN (C1=9 ┆ … ┆ /VIN (C1=5 ┆ /VIN (C1= ┆ /VIN (C1= ┆ /VIN (C1= │
│ ---  ┆ .4e-05,CIN ┆ .4e-05,CIN ┆ .4e-05,CIN ┆   ┆ .7e-05,CIN ┆ 5.7e-05,C ┆ 5.7e-05,C ┆ 5.7e-05,C │
│ f64  ┆ =7e-05,COU ┆ =7e-05,COU ┆ =7e-05,COU ┆   ┆ =1.5e-05,C ┆ IN=1.5e-0 ┆ IN=1.5e-0 ┆ IN=1.5e-0 │
│      ┆ …          ┆ …          ┆ …          ┆   ┆ …          ┆ 5,C…      ┆ 5,C…      ┆ 5,C…      │
│      ┆ ---        ┆ ---        ┆ ---        ┆   ┆ ---        ┆ ---       ┆ ---       ┆ ---       │
│      ┆ i64        ┆ f64        ┆ f64        ┆   ┆ f64        ┆ f64       ┆ f64       ┆ i64       │
╞══════╪════════════╪════════════╪════════════╪═══╪════════════╪═══════════╪═══════════╪═══════════╡
│ 0.0  ┆ 0          ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778   ┆ 0.88889   ┆ 1         │
│ 0.01 ┆ 0          ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778   ┆ 0.88889   ┆ 1         │
│ 0.02 ┆ 0          ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778   ┆ 0.88889   ┆ 1         │
│ 0.03 ┆ 0          ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778   ┆ 0.88889   ┆ 1         │
│ 0.04 ┆ 0          ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778   ┆ 0.88889   ┆ 1         │
└──────┴────────────┴────────────┴────────────┴───┴────────────┴───────────┴───────────┴───────────┘

In [5]:
df_output_1 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Constant_1st_Neuron_Vin0-1_v2.csv"))
df_output_2 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Constant_1st_Neuron_Vin1-2_v2.csv"))
df_output_2.head(50)

shape: (50, 10_000)
┌──────┬────────────┬────────────┬────────────┬───┬────────────┬───────────┬───────────┬───────────┐
│ X    ┆ /Vout_buff ┆ /Vout_buff ┆ /Vout_buff ┆ … ┆ /Vout_buff ┆ /Vout_buf ┆ /Vout_buf ┆ /Vout_buf │
│ ---  ┆ er2 (C1=9. ┆ er2 (C1=9. ┆ er2 (C1=9. ┆   ┆ er2 (C1=5. ┆ fer2 (C1= ┆ fer2 (C1= ┆ fer2 (C1= │
│ f64  ┆ 4e-05,CIN= ┆ 4e-05,CIN= ┆ 4e-05,CIN= ┆   ┆ 7e-05,CIN= ┆ 5.7e-05,C ┆ 5.7e-05,C ┆ 5.7e-05,C │
│      ┆ …          ┆ …          ┆ …          ┆   ┆ …          ┆ IN=…      ┆ IN=…      ┆ IN=…      │
│      ┆ ---        ┆ ---        ┆ ---        ┆   ┆ ---        ┆ ---       ┆ ---       ┆ ---       │
│      ┆ f64        ┆ f64        ┆ f64        ┆   ┆ f64        ┆ f64       ┆ f64       ┆ f64       │
╞══════╪════════════╪════════════╪════════════╪═══╪════════════╪═══════════╪═══════════╪═══════════╡
│ 0.0  ┆ 0.04193    ┆ 0.04193    ┆ 0.04193    ┆ … ┆ 0.026663   ┆ 0.026663  ┆ 0.026663  ┆ 0.026663  │
│ 0.01 ┆ 0.04193    ┆ 0.04193    ┆ 0.04193    ┆ … ┆ 0.026663   ┆ 0.026663  ┆ 0.026663  ┆ 0.026663  │
│ 0.02 ┆ 0.04193    ┆ 0.04193    ┆ 0.04193    ┆ … ┆ 0.026663   ┆ 0.026663  ┆ 0.026663  ┆ 0.026663  │
│ 0.03 ┆ 0.04193    ┆ 0.04193    ┆ 0.04193    ┆ … ┆ 0.026663   ┆ 0.026663  ┆ 0.026663  ┆ 0.026663  │
│ 0.04 ┆ 0.04193    ┆ 0.04193    ┆ 0.04193    ┆ … ┆ 0.026664   ┆ 0.026664  ┆ 0.026664  ┆ 0.026664  │
│ …    ┆ …          ┆ …          ┆ …          ┆ … ┆ …          ┆ …         ┆ …         ┆ …         │
│ 0.45 ┆ 0.041969   ┆ 0.041987   ┆ 0.042001   ┆ … ┆ 0.02952    ┆ 0.029531  ┆ 0.02954   ┆ 0.029548  │
│ 0.46 ┆ 0.041973   ┆ 0.041993   ┆ 0.042009   ┆ … ┆ 0.029759   ┆ 0.029773  ┆ 0.029782  ┆ 0.02979   │
│ 0.47 ┆ 0.041976   ┆ 0.042      ┆ 0.042019   ┆ … ┆ 0.030014   ┆ 0.030027  ┆ 0.030038  ┆ 0.030046  │
│ 0.48 ┆ 0.04198    ┆ 0.042006   ┆ 0.042029   ┆ … ┆ 0.030268   ┆ 0.030282  ┆ 0.030293  ┆ 0.030302  │
│ 0.49 ┆ 0.041984   ┆ 0.042015   ┆ 0.042043   ┆ … ┆ 0.030522   ┆ 0.030537  ┆ 0.030549  ┆ 0.030558  │
└──────┴────────────┴────────────┴────────────┴───┴────────────┴───────────┴───────────┴───────────┘

### 2.1 Goal of Expand Function

The goal with this function is to make is so that the paramaeters of the simulation run that are embedded in the column name are instead added to the tail of the dataframe.
This means afetr using the expand fucntion, each column in the dataframe is one of the simulation run, where each entry is the voltage at a particular time, while at the end of the column the parameetrs are added, for eahc parameetr one new row. 

In [6]:
def expand(df, params=None, id_col="X"):
    if params is None:
        params = ["C1", "CIN", "COUT", "L", "R1", "w1"]

    cols = [c for c in df.columns if c != id_col]

    def parse_col(colname: str):
        try:
            inside = colname.split("(", 1)[1].split(")", 1)[0]
        except Exception:
            return {}
        parsed = {}
        for part in inside.split(","):
            if "=" not in part:
                continue
            k, v = part.split("=", 1)
            try:
                parsed[k.strip()] = float(v.strip())
            except Exception:
                parsed[k.strip()] = None
        return parsed

    parsed = [parse_col(c) for c in cols]

    rows = []
    for p in params:
        row = {id_col: p}
        for i, c in enumerate(cols):
            row[c] = parsed[i].get(p)
        rows.append(row)
    params_df = pl.DataFrame(rows)

    # concat with relaxed typing
    combined = pl.concat([df.with_columns(pl.col(id_col).cast(pl.Utf8)), params_df],
                         how="vertical_relaxed")

    # overwrite the id_col with sequential labels like in pandas
    labels = [str(i) for i in range(len(df))] + params  # 0..N-1, then C1, ...
    combined = combined.with_columns(pl.Series(id_col, labels))

    return combined

In [7]:
%%time

df_input_1_combined = expand(df_input_1[:NUMBER_DATASET_LENGTH])
df_input_2_combined = expand(df_input_2[:NUMBER_DATASET_LENGTH])

CPU times: user 12.3 s, sys: 1.72 s, total: 14 s
Wall time: 11.5 s


In [8]:
df_input_1_combined.head()

shape: (5, 10_000)
┌─────┬────────────┬────────────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ X   ┆ /VIN (C1=9 ┆ /VIN (C1=9 ┆ /VIN (C1=9 ┆ … ┆ /VIN (C1=5 ┆ /VIN (C1=5 ┆ /VIN (C1= ┆ /VIN (C1= │
│ --- ┆ .4e-05,CIN ┆ .4e-05,CIN ┆ .4e-05,CIN ┆   ┆ .7e-05,CIN ┆ .7e-05,CIN ┆ 5.7e-05,C ┆ 5.7e-05,C │
│ str ┆ =7e-05,COU ┆ =7e-05,COU ┆ =7e-05,COU ┆   ┆ =1.5e-05,C ┆ =1.5e-05,C ┆ IN=1.5e-0 ┆ IN=1.5e-0 │
│     ┆ …          ┆ …          ┆ …          ┆   ┆ …          ┆ …          ┆ 5,C…      ┆ 5,C…      │
│     ┆ ---        ┆ ---        ┆ ---        ┆   ┆ ---        ┆ ---        ┆ ---       ┆ ---       │
│     ┆ f64        ┆ f64        ┆ f64        ┆   ┆ f64        ┆ f64        ┆ f64       ┆ f64       │
╞═════╪════════════╪════════════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ 0   ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778    ┆ 0.88889   ┆ 1.0       │
│ 1   ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778    ┆ 0.88889   ┆ 1.0       │
│ 2   ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778    ┆ 0.88889   ┆ 1.0       │
│ 3   ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778    ┆ 0.88889   ┆ 1.0       │
│ 4   ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 0.66667    ┆ 0.77778    ┆ 0.88889   ┆ 1.0       │
└─────┴────────────┴────────────┴────────────┴───┴────────────┴────────────┴───────────┴───────────┘

In [9]:
df_input_1_combined.tail(50)[:, 0:2]

shape: (50, 2)
┌──────┬─────────────────────────────────┐
│ X    ┆ /VIN (C1=9.4e-05,CIN=7e-05,COU… │
│ ---  ┆ ---                             │
│ str  ┆ f64                             │
╞══════╪═════════════════════════════════╡
│ 1457 ┆ 0.0                             │
│ 1458 ┆ 0.0                             │
│ 1459 ┆ 0.0                             │
│ 1460 ┆ 0.0                             │
│ 1461 ┆ 0.0                             │
│ …    ┆ …                               │
│ CIN  ┆ 0.00007                         │
│ COUT ┆ 0.000035                        │
│ L    ┆ 0.000066                        │
│ R1   ┆ 20473.0                         │
│ w1   ┆ 0.000776                        │
└──────┴─────────────────────────────────┘

Then concatenate them side by side, so your final input DataFrame dfin_const contains:
- All signal traces from dfin1 and dfin2
- All the extracted circuit parameters repeated for each corresponding trace  

In [10]:
def safe_concat(dfs, how="horizontal"):
    """
    Concatenate Polars DataFrames horizontally,
    automatically renaming duplicate columns (X, X_2, X_3, ...).
    """
    seen = {}
    renamed = []
    
    for df in dfs:
        new_cols = []
        for c in df.columns:
            if c not in seen:
                seen[c] = 1
                new_cols.append(c)
            else:
                seen[c] += 1
                new_cols.append(f"{c}_{seen[c]}")
        renamed.append(df.rename(dict(zip(df.columns, new_cols))))
    
    return pl.concat(renamed, how=how)

In [11]:
%%time
df_input_dataset_1_concat = safe_concat([df_input_1_combined, df_input_2_combined])

df_input_dataset_1_concat.tail(8)

CPU times: user 422 ms, sys: 26.6 ms, total: 448 ms
Wall time: 449 ms


shape: (8, 20_000)
┌──────┬────────────┬────────────┬────────────┬───┬────────────┬───────────┬───────────┬───────────┐
│ X    ┆ /VIN (C1=9 ┆ /VIN (C1=9 ┆ /VIN (C1=9 ┆ … ┆ /VIN (C1=5 ┆ /VIN (C1= ┆ /VIN (C1= ┆ /VIN (C1= │
│ ---  ┆ .4e-05,CIN ┆ .4e-05,CIN ┆ .4e-05,CIN ┆   ┆ .7e-05,CIN ┆ 5.7e-05,C ┆ 5.7e-05,C ┆ 5.7e-05,C │
│ str  ┆ =7e-05,COU ┆ =7e-05,COU ┆ =7e-05,COU ┆   ┆ =1.5e-05,C ┆ IN=1.5e-0 ┆ IN=1.5e-0 ┆ IN=1.5e-0 │
│      ┆ …          ┆ …          ┆ …          ┆   ┆ …          ┆ 5,C…      ┆ 5,C…      ┆ 5,C…      │
│      ┆ ---        ┆ ---        ┆ ---        ┆   ┆ ---        ┆ ---       ┆ ---       ┆ ---       │
│      ┆ f64        ┆ f64        ┆ f64        ┆   ┆ f64        ┆ f64       ┆ f64       ┆ f64       │
╞══════╪════════════╪════════════╪════════════╪═══╪════════════╪═══════════╪═══════════╪═══════════╡
│ 1499 ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 1.6667     ┆ 1.7778    ┆ 1.8889    ┆ 2.0       │
│ 1500 ┆ 0.0        ┆ 0.22222    ┆ 0.33333    ┆ … ┆ 1.6667     ┆ 1.7778    ┆ 1.8889    ┆ 2.0       │
│ C1   ┆ 0.000094   ┆ 0.000094   ┆ 0.000094   ┆ … ┆ 0.000057   ┆ 0.000057  ┆ 0.000057  ┆ 0.000057  │
│ CIN  ┆ 0.00007    ┆ 0.00007    ┆ 0.00007    ┆ … ┆ 0.000015   ┆ 0.000015  ┆ 0.000015  ┆ 0.000015  │
│ COUT ┆ 0.000035   ┆ 0.000035   ┆ 0.000035   ┆ … ┆ 0.000035   ┆ 0.000035  ┆ 0.000035  ┆ 0.000035  │
│ L    ┆ 0.000066   ┆ 0.000066   ┆ 0.000066   ┆ … ┆ 0.000079   ┆ 0.000079  ┆ 0.000079  ┆ 0.000079  │
│ R1   ┆ 20473.0    ┆ 20473.0    ┆ 20473.0    ┆ … ┆ 36093.0    ┆ 36093.0   ┆ 36093.0   ┆ 36093.0   │
│ w1   ┆ 0.000776   ┆ 0.000776   ┆ 0.000776   ┆ … ┆ 0.000704   ┆ 0.000704  ┆ 0.000704  ┆ 0.000704  │
└──────┴────────────┴────────────┴────────────┴───┴────────────┴───────────┴───────────┴───────────┘

## 3. Processing Outputs of Dataset 1

Now we do the same for the output datasets.

In [12]:
df_output_1 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Constant_1st_Neuron_Vin0-1_v2.csv"))
df_output_2 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Constant_1st_Neuron_Vin1-2_v2.csv"))

In [13]:
df_output_3 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Constant_2nd_Neuron_Vin0-1_v2.csv"))
df_output_4 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Constant_2nd_Neuron_Vin1-2_v2.csv"))

In [14]:
%%time

df_output_1_combined = expand(df_output_1[:NUMBER_DATASET_LENGTH])
df_output_2_combined = expand(df_output_2[:NUMBER_DATASET_LENGTH])
df_output_3_combined = expand(df_output_1[:NUMBER_DATASET_LENGTH])
df_output_4_combined = expand(df_output_2[:NUMBER_DATASET_LENGTH])

CPU times: user 10.4 s, sys: 1.08 s, total: 11.5 s
Wall time: 11.2 s


In [15]:
%%time
df_output_1st_neuron_dataset_1_concat = safe_concat([df_output_1_combined, df_output_2_combined])
df_output_2nd_neuron_dataset_1_concat = safe_concat([df_output_3_combined, df_output_4_combined])

CPU times: user 651 ms, sys: 21.1 ms, total: 672 ms
Wall time: 650 ms


## 4. Processing of Dataset 2

In [16]:
df_input_1 = pl.read_csv(str(DATA_LOCATION + "Input/" + "Processed_sin1_v2.csv"))
df_input_2 = pl.read_csv(str(DATA_LOCATION + "Input/" + "Processed_sin2_v2.csv"))
df_input_3 = pl.read_csv(str(DATA_LOCATION + "Input/" + "Processed_sin3_v2.csv"))

In [17]:
df_output_1 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Sin1_v2.csv"))
df_output_2 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Sin2_v2.csv"))
df_output_3 = pl.read_csv(str(DATA_LOCATION + "/Output/" + "Processed_output_Vout_of_Sin3_v2.csv"))

In [18]:
df_input_1_combined = expand(df_input_1[:NUMBER_DATASET_LENGTH])
df_input_2_combined = expand(df_input_2[:NUMBER_DATASET_LENGTH])
df_input_3_combined = expand(df_input_3[:NUMBER_DATASET_LENGTH])

df_input_dataset_2_concat = safe_concat([df_input_1_combined, df_input_2_combined, df_input_3_combined])

In [19]:
df_output_1_combined = expand(df_output_1[:NUMBER_DATASET_LENGTH])
df_output_2_combined = expand(df_output_2[:NUMBER_DATASET_LENGTH])
df_output_3_combined = expand(df_output_1[:NUMBER_DATASET_LENGTH])

df_output_dataset_2_concat = safe_concat([df_output_1_combined, df_output_2_combined, df_output_3_combined])

## 5. Combining the Datasets into One

In [20]:
# We drop the index and tranform it into a pytorch tensor 

VIN = torch.cat([
    torch.tensor(df_input_dataset_1_concat.drop("X", "X_2").to_numpy().astype("float64")),
    torch.tensor(df_output_1st_neuron_dataset_1_concat.drop("X", "X_2").to_numpy().astype("float64")),
    torch.tensor(df_input_dataset_2_concat.drop("X", "X_2", "X_3").to_numpy().astype("float64"))
], 1)

VOUT = torch.cat([
    torch.tensor(df_output_1st_neuron_dataset_1_concat.drop("X", "X_2").to_numpy().astype("float64")),
    torch.tensor(df_output_2nd_neuron_dataset_1_concat.drop("X", "X_2").to_numpy().astype("float64")),
    torch.tensor(df_output_dataset_2_concat.drop("X", "X_2", "X_3").to_numpy().astype("float64"))
], 1)

VIN.shape, VOUT.shape

(torch.Size([1507, 63993]), torch.Size([1507, 63993]))

In [21]:
X = torch.tensor(VIN.clone().detach()).float().T.unsqueeze(2)
Y = torch.tensor(VOUT.clone().detach()).float().T

X.shape, Y.shape

/tmp/ipykernel_25497/2295200637.py:1: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  X = torch.tensor(VIN.clone().detach()).float().T.unsqueeze(2)
/tmp/ipykernel_25497/2295200637.py:2: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  Y = torch.tensor(VOUT.clone().detach()).float().T


(torch.Size([63993, 1507, 1]), torch.Size([63993, 1507]))

- Randomly shuffle all simulation traces.
- Train, Validation, Test Split

In [22]:
# For small dataset

rand_idx = torch.randperm(X.shape[0])

toss_idx = rand_idx[:int(X.shape[0]*0.5)]
train_idx =rand_idx[int(X.shape[0]*0.5):int(X.shape[0]*0.7)]
valid_idx = rand_idx[int(X.shape[0]*0.8):int(X.shape[0]*0.95)]
test_idx  = rand_idx[int(X.shape[0]*0.95):]

In [22]:
# For large dataset
rand_idx = torch.randperm(X.shape[0])

train_idx = rand_idx[:int(X.shape[0]*0.8)]
valid_idx = rand_idx[int(X.shape[0]*0.8):int(X.shape[0]*0.9)]
test_idx  = rand_idx[int(X.shape[0]*0.9):]

In [23]:
X_train = X[train_idx]
Y_train = Y[train_idx]
X_valid = X[valid_idx]
Y_valid = Y[valid_idx]
X_test  = X[test_idx]
Y_test  = Y[test_idx]

data = {'X_train':X_train, 'Y_train':Y_train,
        'X_valid':X_valid, 'Y_valid':Y_valid,
        'X_test':X_test, 'Y_test':Y_test}

- Normalization + downsampling
- This is a form of dimensionality reduction so the model doesn’t have to process the full 1507-point signal

In [24]:
# :1500:15 means: take columns 0, 15, 30, …, 1495.
# That gives you 100 evenly spaced points across the whole 1500-step window

for k,v in data.items():
    norms = [1.0e-04, 1.0e-04, 1.0e-04, 1.0e-04, 50000, 0.001]
    if "X" in k:
        X1 = v[:,:1500:15,:]
        X2 = v[:,1501:,:]
        for i in range(6):
            X2[:,i,:] /= norms[i]
        X = torch.cat((X1, X2), dim=1)
    else:
        X = v[:,:1500:15]

    data[k] = X

# Simply take the first 100 time steps in full resolution

for k,v in data.items():
    norms = [1.0e-04, 1.0e-04, 1.0e-04, 1.0e-04, 50000, 0.001]
    if "X" in k:
        X1 = v[:,:100,:]
        X2 = v[:,1501:,:]
        for i in range(6):
            X2[:,i,:] /= norms[i]
        X = torch.cat((X1, X2), dim=1)
    else:
        X = v[:,:100]

    data[k] = X

In [25]:
# Check the size for our datasets.
for name, subset in data.items():
    print(name, subset.shape)

X_train torch.Size([12799, 106, 1])
Y_train torch.Size([12799, 100])
X_valid torch.Size([9599, 106, 1])
Y_valid torch.Size([9599, 100])
X_test torch.Size([3200, 106, 1])
Y_test torch.Size([3200, 100])


In [26]:
# Save dataset
# torch.save(data, f'./data/dataset.ds')

# or 

torch.save(data, f'./data/small_dataset.ds')

# 5. Second Dataset: Flexible Neuromorphic Circuit

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import torch
import re
import warnings
import polars as pl

In [1]:
DATA_LOCATION = "data/V3/"

In [4]:
df1 = pl.read_csv(str(DATA_LOCATION + "Power_const_filtered.csv"))
df2 = pl.read_csv(str(DATA_LOCATION + "Power_Sin_filtered.csv"))
df3 = pl.read_csv(str(DATA_LOCATION + "VIN_const_filtered.csv"))
df4 = pl.read_csv(str(DATA_LOCATION + "VIN_Sin_filtered.csv"))
df5 = pl.read_csv(str(DATA_LOCATION + "Vout_const_filtered.csv"))
df6 = pl.read_csv(str(DATA_LOCATION + "Vout_Sin_filtered.csv"))

In [12]:
print(df1.shape)
print(df2.shape)
print(df3.shape)
print(df4.shape)
print(df5.shape)
print(df6.shape)

(10001, 10001)
(10001, 7129)
(10001, 10001)
(10001, 7129)
(10001, 10001)
(10001, 7045)


In [5]:
df3.head()

shape: (5, 10_001)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ /VIN      ┆ /VIN      ┆ /VIN (VIN ┆ /VIN (VIN ┆ … ┆ /VIN (VIN ┆ /VIN (VIN ┆ /VIN      ┆ /VIN (VI │
│ (VIN=0) X ┆ (VIN=0) Y ┆ =0.000300 ┆ =0.000600 ┆   ┆ =2.9994)  ┆ =2.9997)  ┆ (VIN=3) Y ┆ N=1.7980 │
│ ---       ┆ ---       ┆ 03) Y     ┆ 06) Y     ┆   ┆ Y         ┆ Y         ┆ ---       ┆ 8) Y     │
│ f64       ┆ i64       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ i64       ┆ ---      │
│           ┆           ┆ f64       ┆ f64       ┆   ┆ f64       ┆ f64       ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 0.0       ┆ 0         ┆ 0.0       ┆ 0.0       ┆ … ┆ 0.0       ┆ 0.0       ┆ 0         ┆ 0.0      │
│ 0.001     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 2.9994    ┆ 2.9997    ┆ 3         ┆ 1.79808  │
│ 0.002     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 2.9994    ┆ 2.9997    ┆ 3         ┆ 1.79808  │
│ 0.003     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 2.9994    ┆ 2.9997    ┆ 3         ┆ 1.79808  │
│ 0.004     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 2.9994    ┆ 2.9997    ┆ 3         ┆ 1.79808  │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [6]:
df4.head()

shape: (5, 7_129)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ /VIN (a=0 ┆ /VIN (a=0 ┆ /VIN (a=0 ┆ /VIN (a=0 ┆ … ┆ /VIN (a=1 ┆ /VIN (a=1 ┆ /VIN (a=1 ┆ /VIN (a= │
│ .5,b=1,f= ┆ .5,b=1,f= ┆ .5,b=1,f= ┆ .5,b=1,f= ┆   ┆ .5,b=356, ┆ .5,b=356, ┆ .5,b=356, ┆ 1.2,b=27 │
│ 1) X      ┆ 1) Y      ┆ 1.5) Y    ┆ 2) Y      ┆   ┆ f=4) Y    ┆ f=4.5) Y  ┆ f=5) Y    ┆ 6,f=2.5) │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ Y        │
│ f64       ┆ f64       ┆ f64       ┆ f64       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 0.0       ┆ 0.508726  ┆ 0.508726  ┆ 0.508726  ┆ … ┆ 1.395365  ┆ 1.395365  ┆ 1.395365  ┆ 0.006574 │
│ 0.001     ┆ 0.511867  ┆ 0.513437  ┆ 0.515007  ┆ … ┆ 1.433004  ┆ 1.437712  ┆ 1.442421  ┆ 0.008709 │
│ 0.002     ┆ 0.515008  ┆ 0.518147  ┆ 0.521286  ┆ … ┆ 1.470682  ┆ 1.480105  ┆ 1.489529  ┆ 0.011132 │
│ 0.003     ┆ 0.518147  ┆ 0.522856  ┆ 0.527562  ┆ … ┆ 1.508377  ┆ 1.522511  ┆ 1.536643  ┆ 0.013843 │
│ 0.004     ┆ 0.521286  ┆ 0.527562  ┆ 0.533833  ┆ … ┆ 1.546065  ┆ 1.564898  ┆ 1.583719  ┆ 0.016851 │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [7]:
def safe_concat(dfs, how="horizontal"):
    """
    Concatenate Polars DataFrames horizontally,
    automatically renaming duplicate columns (X, X_2, X_3, ...).
    """
    seen = {}
    renamed = []
    
    for df in dfs:
        new_cols = []
        for c in df.columns:
            if c not in seen:
                seen[c] = 1
                new_cols.append(c)
            else:
                seen[c] += 1
                new_cols.append(f"{c}_{seen[c]}")
        renamed.append(df.rename(dict(zip(df.columns, new_cols))))
    
    return pl.concat(renamed, how=how)

In [9]:
%%time
df_input_concat = safe_concat([df3, df4])
df_input_concat.head()

CPU times: user 42.4 ms, sys: 1.97 ms, total: 44.3 ms
Wall time: 42.5 ms


shape: (5, 17_130)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ /VIN      ┆ /VIN      ┆ /VIN (VIN ┆ /VIN (VIN ┆ … ┆ /VIN (a=1 ┆ /VIN (a=1 ┆ /VIN (a=1 ┆ /VIN (a= │
│ (VIN=0) X ┆ (VIN=0) Y ┆ =0.000300 ┆ =0.000600 ┆   ┆ .5,b=356, ┆ .5,b=356, ┆ .5,b=356, ┆ 1.2,b=27 │
│ ---       ┆ ---       ┆ 03) Y     ┆ 06) Y     ┆   ┆ f=4) Y    ┆ f=4.5) Y  ┆ f=5) Y    ┆ 6,f=2.5) │
│ f64       ┆ i64       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ Y        │
│           ┆           ┆ f64       ┆ f64       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 0.0       ┆ 0         ┆ 0.0       ┆ 0.0       ┆ … ┆ 1.395365  ┆ 1.395365  ┆ 1.395365  ┆ 0.006574 │
│ 0.001     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 1.433004  ┆ 1.437712  ┆ 1.442421  ┆ 0.008709 │
│ 0.002     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 1.470682  ┆ 1.480105  ┆ 1.489529  ┆ 0.011132 │
│ 0.003     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 1.508377  ┆ 1.522511  ┆ 1.536643  ┆ 0.013843 │
│ 0.004     ┆ 0         ┆ 0.0003    ┆ 0.0006    ┆ … ┆ 1.546065  ┆ 1.564898  ┆ 1.583719  ┆ 0.016851 │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [10]:
%%time
df_output_concat = safe_concat([df5, df6])
df_output_concat.head()

CPU times: user 57 ms, sys: 6.17 ms, total: 63.1 ms
Wall time: 61.4 ms


shape: (5, 17_046)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ /Vout_buf ┆ /Vout_buf ┆ /Vout_buf ┆ /Vout_buf ┆ … ┆ /Vout_buf ┆ /Vout_buf ┆ /Vout_buf ┆ /Vout_bu │
│ fer2      ┆ fer2      ┆ fer2 (VIN ┆ fer2 (VIN ┆   ┆ fer2 (a=1 ┆ fer2 (a=1 ┆ fer2 (a=1 ┆ ffer2    │
│ (VIN=0) X ┆ (VIN=0) Y ┆ =0.000300 ┆ =0.000600 ┆   ┆ .5,b=311, ┆ .5,b=311, ┆ .5,b=311, ┆ (a=1,b=4 │
│ ---       ┆ ---       ┆ 03)…      ┆ 06)…      ┆   ┆ f=2…      ┆ f=3…      ┆ f=3…      ┆ 6,f=2) Y │
│ f64       ┆ f64       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│           ┆           ┆ f64       ┆ f64       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ 0.001     ┆ 0.288568  ┆ 0.288568  ┆ 0.288568  ┆ … ┆ 0.288568  ┆ 0.288568  ┆ 0.288568  ┆ 0.288568 │
│ 0.002     ┆ 0.288569  ┆ 0.288569  ┆ 0.288569  ┆ … ┆ 0.288569  ┆ 0.288569  ┆ 0.288569  ┆ 0.288569 │
│ 0.003     ┆ 0.28857   ┆ 0.28857   ┆ 0.28857   ┆ … ┆ 0.28857   ┆ 0.28857   ┆ 0.28857   ┆ 0.28857  │
│ 0.004     ┆ 0.288573  ┆ 0.288573  ┆ 0.288573  ┆ … ┆ 0.288573  ┆ 0.288573  ┆ 0.288573  ┆ 0.288573 │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [11]:
print(df_input_concat.shape)
print(df_output_concat.shape)

(10001, 17130)
(10001, 17046)
